### ISSUE TITLE
Relative go-to line

### ISSUE THREAD

**Issue by SplittyDev** [microsoft/vscode#65539]
I frequently use the go-to line (`Ctrl+G`) feature to jump to specific lines, and I'd love to see support for relative line numbers in the future.

Relative line-jumping could be added by parsing signs at the beginning of the command:

| Command      | Action
| ------------ | ------------------
| `Ctrl+G`, `n`  | Jump to line `n`
| `Ctrl+G`, `+n` | Jump to line `min(line_count, current_line + n)`
| `Ctrl+G`, `-n` | Jump to line `max(1, current_line - n)`

Currently, the `+` sign is ignored, the `-` sign produces a nonsensical output (`"Go to line -20"`) and jumps to line 1. I believe that parsing of signs wouldn't break anyone's workflow in any way.

**Comment by usernamehw** [comment]
https://github.com/Microsoft/vscode/issues/14162

**Comment by SplittyDev** [comment]
@usernamehw This was two years ago, might wanna give it another chance.
It's not a feature that bloats the editor in any way, imho it doesn't interfere with the objective of keeping VSCode lean and fast.

Of course, that's not for me to decide, but I think it would be nice to give it a chance and think about it.

**Comment by thecakeisalie25** [comment]
Couldn't this be accomplished with an extension?

**Comment by SplittyDev** [comment]
@thecakeisalie25 I'm sure it could, I just think it should be part of the core functionality. The current behavior is kinda broken, it would be a small addition that wouldn't cause any problems and it would replace a currently broken mechanism (signed numbers in the go-to box) with a functional one.

**Comment by thecakeisalie25** [comment]
@SplittyDev Fair. And it shouldn't be that hard to implement:
```js
if(goto.startsWith("+")
{
    line+=goto.toInt();
}
else if(gotoInput.startsWith("-")
{
    line-=goto.toInt();
}
else
{
    currentGotoMethod(goto);
}
```

**Comment by NikxDa** [comment]
@thecakeisalie25 Even easier:
`if (goto.startsWith ("+") || goto.startsWith ("-")) goto = currentLine + Number (goto)
currentGotoMethod (goto);`

**Comment by tombresson** [comment]
In case anyone is looking for an extension that implements this functionality, [this one](https://marketplace.visualstudio.com/items?itemName=retepaskab.relativegoto&ssr=false#overview) has been released for a few months. Seems to implement the missing feature, albeit not as seamless as a built-in solution could.

**Comment by bpasero** [comment]
I am not seeing a good way to implement this into core right now because a while ago we added support for negative line numbers that would navigate from the end of the file. We would have to find another syntax to support relative navigation or change this support which may be surprising.

Since there is an extension I would tend to mark this as closed, but give it a couple of days for feedback.

**Comment by SplittyDev** [comment]
I guess one way would be `~N` and `~-N`

Examples:
```
~20 // jumps 20 lines ahead
~-20 // jumps 20 lines behind
```

I understand it though if this isn't a priority in any way.
Not sure if people would actually use this, would be nice go gather some more feedback

**Comment by bpasero** [comment]
I think this is not easily discoverable and would prefer an extra command to trigger this as the extension suggests.

**Comment by NikxDa** [comment]
I second the closing. With an extension in place & the syntax being occupied in the meantime, I don't think there's a nice way for this to go forward in core. If extension installations are an indicator, the feature isn't hugely popular either.

**Comment by Jaimeloeuf** [comment]
> In case anyone is looking for an extension that implements this functionality, [this one](https://marketplace.visualstudio.com/items?itemName=retepaskab.relativegoto&ssr=false#overview) has been released for a few months. Seems to implement the missing feature, albeit not as seamless as a built-in solution could.

Just for quick reference if anyone comes across this after searching, the extension mentioned is 2 years plus old and has never been updated. There is a bug with it when jumping lines when you have a text selection. It's original source code is also not available, which makes it impossible to build on top off / contribute a change, thus I made an alternative to it!

https://marketplace.visualstudio.com/items?itemName=EnkelDigital.relative-goto

Do check it out, this extension is open source on github and you are free to fork / contribute to it!  
It also contains more than just jumping using relative lines, it also allows you to select a text between your original cursor position to the final jump destination.

**Comment by DanielBreiner** [comment]
The chain of abandoning extensions continues 😄. The one mentioned above was missing some features and hasn't been updated in 4 years - I created a new one similar to it, but tried keeping the command very similar to the built-in "Go to Line/Character...". It also has:
- automatic switching to relative line numbers and back
- preview highlight of the destination line/character/selection
- a separate command to prefill the input with "-" to make jumping up faster (default ctrl+g and ctrl+shift+g)

https://marketplace.visualstudio.com/items?itemName=DanielBreiner.go-to-relative

![showcase](https://github.com/microsoft/vscode/assets/43221175/8c73ea97-986f-477d-b88d-95befc7e3cad)

**Comment by Jaimeloeuf** [comment]
> > In case anyone is looking for an extension that implements this functionality, [this one](https://marketplace.visualstudio.com/items?itemName=retepaskab.relativegoto&ssr=false#overview) has been released for a few months. Seems to implement the missing feature, albeit not as seamless as a built-in solution could.
> 
> Just for quick reference if anyone comes across this after searching, the extension mentioned is 2 years plus old and has never been updated. There is a bug with it when jumping lines when you have a text selection. It's original source code is also not available, which makes it impossible to build on top off / contribute a change, thus I made an alternative to it!
> 
> https://marketplace.visualstudio.com/items?itemName=EnkelDigital.relative-goto
> 
> Do check it out, this extension is open source on github and you are free to fork / contribute to it! It also contains more than just jumping using relative lines, it also allows you to select a text between your original cursor position to the final jump destination.


Hey @DanielBreiner ! Thanks alot for creating a new one, life got in the way and just didn't follow up with anything, so really appreciate you doing the good work!

Some context for everyone else that will ever come across this thread.

I was actually able to make quite a few improvements to the extension after the one that is available on the marketplace, which I use as a local plugin. I am now unable to push any new changes to the marketplace or push new ones to the same organisation due to an azure outage in the past. I am now completely locked out of my azure account and Organisation which I used to publish this and other extensions, and have no way of getting it back despite contacting azure support for quite abit.

That and life just got in the way and didn't realise how many people actually use this plugin until I just went to check!

I spent almost a year sometime in 2022 ish using neovim and the awesome [leap.nvim](https://github.com/ggandor/leap.nvim) plugin which allows you to jump to words directly without requiring me to jump to a line before navigating to the position.

I got so used to the workflow that when I came back to VS Code, I found and now fully adopt the use of [this extension](https://marketplace.visualstudio.com/items?itemName=shadowndacorner.vscode-easymotion). Although there are still times where I still use this extension for things like highlight over a few lines to cut, it is becoming very rare as all my "movements" now use the vscode-easymotion plugin.

Thanks for reading and I hope you give the vscode-easymotion plugin a try too! Might improve your workflow lots too!
